apis/privilegedaccessmanager/v1alpha1/types.generated.go

```diff
@@ -0,0 +1,195 @@
+//
+// A single user might be part of the 'approvers' ACL for multiple steps in this
+// workflow, but they can only approve once and that approval is only considered
+// to satisfy the approval step at which it was granted.
+// +kcc:proto=google.cloud.privilegedaccessmanager.v1.ManualApprovals
+type ManualApprovals struct {
+	// Optional. Whether the approvers need to provide a justification for their
+	// actions.
+	// +optional
+	RequireApproverJustification *bool `json:"requireApproverJustification,omitempty"`
+
+	// Optional. List of approval steps in this workflow. These steps are followed
+	// in the specified order sequentially. Only 1 step is supported.
+	// +optional
+	Steps []*Step `json:"step,omitempty"`
+}
+
+// Step represents a logical step in a manual approval workflow.
+// +kcc:proto=google.cloud.privilegedaccessmanager.v1.ManualApprovals.Step
+type Step struct {
+	// Optional. The potential set of approvers in this step. This list must
+	// contain at most one entry.
+	// +optional
+	Approvers []*AccessControlEntry `json:"approvers,omitempty"`
+
+	// Required. How many users from the above list need to approve. If there
+	// aren't enough distinct users in the list, then the workflow indefinitely
+	// blocks. Should always be greater than 0. 1 is the only supported value.
+	// +required
+	ApprovalsNeeded *int32 `json:"approvalsNeeded,omitempty"`
+
+	// Optional. Additional email addresses to be notified when a grant is
+	// pending approval.
+	// +optional
+	ApproverEmailRecipients []string `json:"approverEmailRecipients,omitempty"`
+}
+
+// Privileged access that this service can be used to gate.
+// +kcc:proto=google.cloud.privilegedaccessmanager.v1.PrivilegedAccess
+type PrivilegedAccess struct {
+	// Access to a Google Cloud resource through IAM.
+	// +required
+	GcpIAMAccess *GcpIamAccess `json:"gcpIAMAccess,omitempty"`
+}
+
+// GcpIamAccess represents IAM based access control on a Google Cloud
+// resource. Refer to https://cloud.google.com/iam/docs to understand more
+// about IAM.
+// +kcc:proto=google.cloud.privilegedaccessmanager.v1.PrivilegedAccess.GcpIamAccess
+type GcpIamAccess struct {
+	// Required. The type of this resource.
+	// +required
+	ResourceType *string `json:"resourceType,omitempty"`
+
+	// The Project that this privileged access is granted to.
+	// One and only one of 'projectRef', 'folderRef', or 'organizationRef' must
+	// be set.
+	// +optional
+	ProjectRef *refs.ProjectRef `json:"projectRef,omitempty"`
+
+	// The Folder that this privileged access is granted to.
+	// One and only one of 'projectRef', 'folderRef', or 'organizationRef' must
+	// be set.
+	// +optional
+	FolderRef *refs.FolderRef `json:"folderRef,omitempty"`
+
+	// The Organization that this privileged access is granted to.
+	// One and only one of 'projectRef', 'folderRef', or 'organizationRef' must
+	// be set.
+	// +optional
+	OrganizationRef *refs.OrganizationRef `json:"organizationRef,omitempty"`
```

Comment:
Do you know what is the relationship between `spec.projectRef` and `spec. privilegedAccess. gcpIAMAccess.projectRef`?

Reply:
Right now they must be the same, but they also must both be specified.

I was thinking of hiding this field and generate the value for the users, but I'm concerned that the service team extends this field's behavior unexpectedly so decided to expose the exact same interface as the underlying GCP API.

Overall this resource was designed weirdly.

Reply:
I see. Thanks for the explanation.